compiler/optimizer/demand.go

```diff
@@ -0,0 +1,198 @@
+package optimizer
+
+import (
+	//"fmt"
+	"github.com/brimdata/zed/compiler/ast/dag"
+)
+
+type DemandAll struct{}
+type DemandKeys map[string]Demand // No empty values.
+
+type Demand interface {
+	isDemand()
+}
+
+func (demand DemandAll) isDemand()  {}
+func (demand DemandKeys) isDemand() {}
+
+func demandNone() Demand {
+	return DemandKeys(make(map[string]Demand, 0))
+}
+
+func demandIsValid(demand Demand) bool {
+	switch demand := demand.(type) {
+	case DemandAll:
+		return true
+	case DemandKeys:
+		for _, v := range demand {
+			if !demandIsValid(v) || demandIsEmpty(v) {
+				return false
+			}
+		}
+		return true
+	default:
+		panic("Unreachable")
+	}
+}
+
+func demandIsEmpty(demand Demand) bool {
+	switch demand := demand.(type) {
+	case DemandAll:
+		return false
+	case DemandKeys:
+		return len(demand) == 0
+	default:
+		panic("Unreachable")
+	}
+}
+
+func demandKey(key string, value Demand) Demand {
+	if demandIsEmpty(value) {
+		return value
+	}
+	demand := DemandKeys(make(map[string]Demand, 1))
+	demand[key] = value
+	return demand
+}
+
+func demandUnion(a Demand, b Demand) Demand {
+	if _, ok := a.(DemandAll); ok {
+		return a
+	}
+	if _, ok := b.(DemandAll); ok {
+		return b
+	}
+
+	{
+		a := a.(DemandKeys)
+		b := b.(DemandKeys)
+
+		demand := DemandKeys(make(map[string]Demand, len(a)+len(b)))
+		for k, v := range a {
+			demand[k] = v
+		}
+		for k, v := range b {
+			if v2, ok := a[k]; ok {
+				demand[k] = demandUnion(v, v2)
+			} else {
+				demand[k] = v
+			}
+		}
+		return demand
+	}
+}
+
+func demandForSeq(seq dag.Seq) map[dag.Op]Demand {
+	demands := make(map[dag.Op]Demand)
+	demandForSeqInto(demands, DemandAll{}, seq)
+
+	//walk(seq, true, func(seq dag.Seq) dag.Seq {
+	//    for i := range seq {
+	//        fmt.Println(seq[i], " ", demands[&seq[i]])
+	//    }
+	//    return seq
+	//})
+
+	for _, demand := range demands {
+		if !demandIsValid(demand) {
+			panic("Invalid demand")
+		}
+	}
+	return demands
+}
+
+func demandForSeqInto(demands map[dag.Op]Demand, demandOnSeq Demand, seq dag.Seq) {
+	var demand = demandOnSeq
```

Comment:
Nit: We generally reserve `var` for declaring zero-valued variables.
```suggestion
	demand := demandOnSeq
```